File: apps/badminton-coach/analyze/badminton_coach/session.py
"""Glue: raw pose frames in, conditioned metrics and detected strokes out.

The same class runs over a whole clip here and, in its JavaScript twin
(``web/js/core/session.js``), live in the browser -- so the numbers shown on
court during practice are the numbers in the report afterwards.
"""

from __future__ import annotations

import math
from dataclasses import dataclass, field

from . import landmarks as L
from .biomech import body_scale, core_visibility, frame_metrics, world_to_up_frame
from .filters import LandmarkFilter, RunningMedian
from .strokes import ShotThresholds, StrokeOptions, detect_strokes, summarise_strokes, wrist_speed_series


@dataclass
class SessionOptions:
    racket_arm: str = "right"
    world_filter: dict = field(default_factory=lambda: {"min_cutoff": 1.5, "beta": 0.35})
    """1-Euro settings for the metric world landmarks."""

    image_filter: dict = field(default_factory=lambda: {"min_cutoff": 1.0, "beta": 0.15})
    """Smoother settings for the on-screen image landmarks: the overlay is judged
    by eye, and a jittery skeleton looks broken."""

    scale_window: int = 45
    """Window, in frames, of the running median used for body scale."""

    min_visibility: float = 0.35


def detect_racket_arm(frames) -> dict | None:
    """Guess which hand holds the racket from which wrist swings harder."""
    if len(frames) < 10:
        return None
    trunks = [f["metrics"]["trunk_length"] for f in frames]

    def score(arm: str) -> float:
        speeds = [v for v in wrist_speed_series(frames, arm, trunks)
                  if v is not None and math.isfinite(v)]
        if len(speeds) < 5:
            return -math.inf
        # The 95th percentile, not the mean: the non-racket arm also moves a lot
        # in badminton (it balances and points), but only the racket arm is whipped.
        speeds.sort()
        return speeds[int(0.95 * (len(speeds) - 1))]

    right, left = score("right"), score("left")
    if not math.isfinite(right) or not math.isfinite(left):
        return None
    margin = abs(right - left) / max(right, left, 1e-6)
    return {"arm": "right" if right >= left else "left",
            "margin": margin, "right": right, "left": left}


class PoseSession:
    def __init__(self, options: SessionOptions | None = None):
        self.options = options or SessionOptions()
        self.world_filter = LandmarkFilter(L.COUNT, 3, **self.options.world_filter)
        self.image_filter = LandmarkFilter(L.COUNT, 2, **self.options.image_filter)
        self.trunk_median = RunningMedian(self.options.scale_window)
        self.arm_median = RunningMedian(self.options.scale_window)
        self.frames: list[dict] = []
        self.dropped = 0

    @property
    def racket_arm(self) -> str:
        return self.options.racket_arm

    def set_racket_arm(self, arm: str) -> None:
        """Change the racket hand, re-measuring every stored frame.

        A mid-session correction fixes the history too, rather than leaving a seam
        in the data.
        """
        if arm not in ("left", "right"):
            raise ValueError(f"bad racket arm: {arm}")
        if arm == self.options.racket_arm:
            return
        self.options.racket_arm = arm
        for f in self.frames:
            f["metrics"] = frame_metrics(
                f["world"], arm,
                trunk=f["metrics"]["trunk_length"],
                arm_length=f["metrics"]["arm_length"],
            )

    def push(self, t: float, world, image=None, frame=None) -> dict | None:
        """Add one detected pose. ``world`` is raw MediaPipe (y-down)."""
        if not world or len(world) < L.COUNT:
            return None
        if image and core_visibility(image) < self.options.min_visibility:
            self.dropped += 1
            return None

        filtered = self.world_filter.filter(world_to_up_frame(world), t)
        filtered_image = self.image_filter.filter(image, t) if image else None

        raw = body_scale(filtered)
        trunk = self.trunk_median.push(raw.trunk)
        arm_raw = raw.arm_right if self.racket_arm == "right" else raw.arm_left
        arm_length = self.arm_median.push(arm_raw)

        analysed = {
            "t": t,
            "frame": frame,
            "world": filtered,
            "image": filtered_image,
            "metrics": frame_metrics(filtered, self.racket_arm, trunk=trunk, arm_length=arm_length),
        }
        self.frames.append(analysed)
        return analysed

    def push_track(self, track: dict) -> "PoseSession":
        """Load a whole track written by :mod:`badminton_coach.extract`."""
        for f in track["frames"]:
            self.push(f["t"], f["world"], f.get("image"), f.get("frame"))
        return self

    def strokes(self, options: StrokeOptions | None = None,
                thresholds: ShotThresholds | None = None) -> list[dict]:
        return detect_strokes(self.frames, options, thresholds)

    def analyse(self, options: StrokeOptions | None = None,
                thresholds: ShotThresholds | None = None) -> dict:
        strokes = self.strokes(options, thresholds)
        return {
            "racket_arm": self.racket_arm,
            "frames": len(self.frames),
            "dropped": self.dropped,
            "duration": (self.frames[-1]["t"] - self.frames[0]["t"]) if self.frames else 0.0,
            "strokes": strokes,
            "summary": summarise_strokes(strokes),
        }

File: apps/badminton-coach/web/js/core/session.js
/**
 * Glue: raw pose frames in, conditioned metrics and detected strokes out.
 *
 * The same class runs live in the browser and offline over a whole clip, so the
 * numbers shown on court during practice are the numbers in the report
 * afterwards. Live use calls {@link PoseSession#push} per frame and
 * {@link PoseSession#recentStrokes}; batch use pushes everything and calls
 * {@link PoseSession#analyse}.
 */

import { LandmarkFilter, RunningMedian } from './filters.js';
import { worldToUpFrame, frameMetrics, bodyScale } from './biomech.js';
import { LANDMARK_COUNT, CORE, side, LM } from './landmarks.js';
import { detectStrokes, summariseStrokes, wristSpeedSeries } from './strokes.js';
import { distance, midpoint, sub } from './vec3.js';
import { coreVisibility } from './biomech.js';

export const DEFAULT_SESSION_OPTIONS = {
  racketArm: 'right',
  /** 1-Euro settings for the metric world landmarks. */
  worldFilter: { minCutoff: 1.5, beta: 0.35 },
  /** 1-Euro settings for the on-screen image landmarks; smoother, since the
   *  overlay is judged by eye and a jittery skeleton looks broken. */
  imageFilter: { minCutoff: 1.0, beta: 0.15 },
  /** Window, in frames, of the running median used for body scale. */
  scaleWindow: 45,
  /** Frames kept in memory. At 30 fps this is two minutes; live sessions do not
   *  need more, and it bounds memory on a phone. */
  maxFrames: 3600,
  /** Drop frames the pose model could barely see. */
  minVisibility: 0.35,
};

/** Guess which hand holds the racket from which wrist swings harder. */
export function detectRacketArm(frames) {
  if (frames.length < 10) return null;
  const trunks = frames.map((f) => f.metrics?.trunkLength ?? NaN);
  const score = (arm) => {
    const speeds = wristSpeedSeries(frames, arm, trunks).filter(Number.isFinite);
    if (speeds.length < 5) return -Infinity;
    // The 95th percentile, not the mean: the non-racket arm also moves a lot in
    // badminton (it balances and points), but only the racket arm gets whipped.
    const sorted = speeds.slice().sort((a, b) => a - b);
    return sorted[Math.floor(0.95 * (sorted.length - 1))];
  };
  const right = score('right');
  const left = score('left');
  if (!Number.isFinite(right) || !Number.isFinite(left)) return null;
  const margin = Math.abs(right - left) / Math.max(right, left, 1e-6);
  return { arm: right >= left ? 'right' : 'left', margin, right, left };
}

export class PoseSession {
  constructor(options = {}) {
    this.options = { ...DEFAULT_SESSION_OPTIONS, ...options };
    this.worldFilter = new LandmarkFilter(LANDMARK_COUNT, 3, this.options.worldFilter);
    this.imageFilter = new LandmarkFilter(LANDMARK_COUNT, 2, this.options.imageFilter);
    this.trunkMedian = new RunningMedian(this.options.scaleWindow);
    this.armMedian = new RunningMedian(this.options.scaleWindow);
    this.frames = [];
    this.dropped = 0;
    this._strokeCount = 0;
  }

  get racketArm() {
    return this.options.racketArm;
  }

  /**
   * Change which hand is treated as the racket hand.
   *
   * Every stored frame is re-measured, so a mid-session correction fixes the
   * history too rather than leaving a seam in the data.
   */
  setRacketArm(arm) {
    if (arm !== 'left' && arm !== 'right') throw new Error(`bad racket arm: ${arm}`);
    if (arm === this.options.racketArm) return;
    this.options.racketArm = arm;
    for (const f of this.frames) {
      f.metrics = frameMetrics(f.world, {
        racketArm: arm,
        trunk: f.metrics.trunkLength,
        armLength: f.metrics.armLength,
      });
    }
  }

  reset() {
    this.worldFilter.reset();
    this.imageFilter.reset();
    this.trunkMedian.reset();
    this.armMedian.reset();
    this.frames = [];
    this.dropped = 0;
    this._strokeCount = 0;
  }

  /**
   * Add one detected pose.
   *
   * @param {object} sample
   * @param {number} sample.t seconds
   * @param {number[][]} sample.world raw MediaPipe world landmarks (y-down)
   * @param {number[][]} [sample.image] normalised image landmarks with visibility
   * @param {number} [sample.frame] source frame number, for reporting
   * @returns {object|null} the analysed frame, or null if it was too poor to use
   */
  push({ t, world, image = null, frame = null }) {
    if (!world || world.length < LANDMARK_COUNT) return null;
    if (image && coreVisibility(image, CORE) < this.options.minVisibility) {
      this.dropped += 1;
      return null;
    }

    const up = worldToUpFrame(world);
    const filtered = this.worldFilter.filter(up, t);
    const filteredImage = image ? this.imageFilter.filter(image, t) : null;

    const raw = bodyScale(filtered);
    const trunk = this.trunkMedian.push(raw.trunk);
    const armRaw = this.options.racketArm === 'right' ? raw.armRight : raw.armLeft;
    const armLength = this.armMedian.push(armRaw);

    const metrics = frameMetrics(filtered, {
      racketArm: this.options.racketArm,
      trunk,
      armLength,
    });

    const analysed = { t, frame, world: filtered, image: filteredImage, metrics };
    this.frames.push(analysed);
    if (this.frames.length > this.options.maxFrames) this.frames.shift();
    return analysed;
  }

  /** Every stroke in the buffer. */
  strokes(options = {}) {
    return detectStrokes(this.frames, { ...options, racketArm: this.racketArm });
  }

  /**
   * Strokes found since the last call -- the live path.
   *
   * The most recent frames are held back by `settle` seconds, because a peak
   * cannot be confirmed as a peak until the frames after it have arrived.
   */
  recentStrokes({ settle = 0.25, ...options } = {}) {
    if (!this.frames.length) return [];
    const cutoff = this.frames[this.frames.length - 1].t - settle;
    const settled = this.frames.filter((f) => f.t <= cutoff);
    const all = detectStrokes(settled, options);
    const fresh = all.slice(this._strokeCount);
    this._strokeCount = all.length;
    return fresh;
  }

  /** Full analysis of everything buffered. */
  analyse(options = {}) {
    const strokes = this.strokes(options);
    return {
      racketArm: this.racketArm,
      frames: this.frames.length,
      dropped: this.dropped,
      duration: this.frames.length
        ? this.frames[this.frames.length - 1].t - this.frames[0].t
        : 0,
      strokes,
      summary: summariseStrokes(strokes),
    };
  }

  /** Latest frame's metrics, for the live readout. */
  get latest() {
    return this.frames.length ? this.frames[this.frames.length - 1] : null;
  }

  /**
   * Current racket-wrist speed, in trunk lengths per second.
   *
   * Computed over the last few frames only, so the live readout costs nothing
   * even when the session buffer holds two minutes of play.
   */
  speedNow(window = 5) {
    const tail = this.frames.slice(-window);
    if (tail.length < 3) return NaN;
    const speeds = wristSpeedSeries(
      tail,
      this.racketArm,
      tail.map((f) => f.metrics.trunkLength),
    ).filter(Number.isFinite);
    return speeds.length ? speeds[speeds.length - 1] : NaN;
  }
}
